File: Problem-of-the-day/16.03.2023/Solution_Adi.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        # base case
        if not inorder or not postorder:
            return None

        # find the root node in the postorder traversal array
        root_val = postorder[-1]
        root = TreeNode(root_val)

        # find the index of the root node in the inorder traversal array
        root_idx = inorder.index(root_val)

        # recursively construct the left and right subtrees
        root.left = self.buildTree(inorder[:root_idx], postorder[:root_idx])
        root.right = self.buildTree(
            inorder[root_idx+1:], postorder[root_idx:-1])

        return root

File: Problem-of-the-day/15.03.2023/Solution_Adi.cpp
class Solution
{
public:
    bool isCompleteTree(TreeNode *root)
    {
        queue<TreeNode *> q;
        q.push(root);
        int i = 0, f = 0;
        while (!q.empty())
        {
            auto node = q.front();
            q.pop();
            if (f && node != NULL)
                return false;
            if (node == NULL)
            {
                f = 1;
                continue;
            }
            q.push(node->left);
            q.push(node->right);
        }
        return true;
    }
};